keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -0,0 +1,730 @@
+6. `anonymousAccessMode`: a string specifying *uncredentialed* access to the Bucket.  This is applicable to cases where the bucket objects are intended to be publicly readable and/or writable. One of:
+   - "private": Default, disallow uncredentialed access to the backend storage.
+   - "publicReadOnly": Read only, uncredentialed users can call ListBucket and GetObject.
+   - "publicWriteOnly": Write only, uncredentialed users can only call PutObject.
+   - "publicReadWrite": Read/Write, same as `ro` with the addition of PutObject being allowed.
+7. `bucketClassName`: (optional for greenfield, omitted for brownfield) name of the associated bucket class. 
+8. `bucketRequest`: an `objectReference` containing the name, namespace and UID of the associated `BucketRequest`.
+9. `allowedNamespaces`: a list of namespaces that are permitted to either create new buckets or to access existing buckets.
+10. `protocol`: protocol-specific field the application will use to access the backend storage.
+     - `name`: supported protocols are: "s3", "gs", and "azureBlob".
+     - `version`: version being used by the backend storage.
+     - `azureBlob`: data required to target a provisioned azure container.
+     - `gs`: data required to target a provisioned GS bucket.
+     - `s3`: data required to target a provisioned S3 bucket.
+11. `parameters`: a copy of the BucketClass parameters.
+12. `bucketAvailable`: if true the bucket has been provisioned. If false then the bucket has not been provisioned and is unable to be accessed.
+
+#### BucketClass
+
+A cluster-scoped custom resource to provide admins control over the handling of bucket provisioning.  The `BucketClass` (BC) defines a retention policy, driver specific parameters, and the provisioner name. A list of allowed namespaces can be specified to restrict new bucket creation and access to existing buckets. A default bucket class can be defined for each supported protocol. This allows the bucket class to be omitted from a `BucketRequest`. All of the `BucketClass` fields are copied to the `Bucket` instance.
+
+> Note: like storage classes, a `BucketClass` is immutable, but a `Bucket` is mutable and contains all of the BC fields.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketClass
+metadata:
+  name: 
+provisioner: [1]
+isDefaultBucketClass: [2]
+protocol: {"azureblob", "gs", "s3", ... } [3]
+anonymousAccessMode: [4]
+retentionPolicy: {"Delete", "Retain"} [5]
+allowedNamespaces: [6]
+  - name:
+parameters: [7]
+```
+
+1. `provisioner`: (required) the name of the vendor-specific driver supporting the `protocol`.
+2. `isDefaultBucketClass`: (optional) boolean, default is false. If set to true then a `BucketRequest` may not need to specify a `BucketClass`. If the greenfield `BucketRequest` omits the `BucketClass` and a default `BucketClass`'s protocol matches the `BucketRequest`'s protocol then the default bucket class is used; otherwise an error is logged.
```

Comment:
Define what happens if for a given protocol, more than one BucketClass is marked as default (on StorageClass we fail provisioning of PVC with no StorageClass in that case).

Reply:
Storageclasses have (or had) this same issue. We have 2 choices: 1) create a BC admission controller that  rejects a new default BC if it matches an existing default BC for the given protocol, 2) let the result be undefined, meaning the first BC retrieved that matches the BR's protocol wins.